corehq/apps/registration/utils.py

```diff
@@ -92,27 +93,36 @@ def safe_unsubscribe_user_from_mailchimp_list(user, list_id, email=None):
         logging.error(e.message)
 
 
-def handle_changed_mailchimp_email(user, old_email, new_email, list_id):
-    def is_user_subscribed_with_email(couch_user):
-        return (couch_user.subscribed_to_commcare_users
-                and couch_user.email == old_email)
-    users_subscribed_with_old_email = [
-        other_user
-        for other_user in CouchUser.all()
-        if is_user_subscribed_with_email(other_user)
-    ]
+def handle_changed_mailchimp_email(user, old_email, new_email):
+    """
+        Checks whether there are other users with old_email who are also subscribed to any mailchimp lists.
+        If not, it safely unsubscribes that email from mailchimp. Then, adds new_email to mailchimp.
+    """
+    def subscribed(couch_user):
+        return couch_user.subscribed_to_commcare_users or not couch_user.email_opt_out
+
+    users_with_old_email = User.objects.filter(email=old_email).values_list('username', flat=True)
+
+    couch_users = CouchUser.view('users/by_username',
+                                 keys=list(users_with_old_email),
+                                 include_docs=True,
+                                 reduce=False,
+                                 ).all()
+
+    users_subscribed_with_old_email = [couch_user.get_id for couch_user in couch_users if subscribed(couch_user)]
+
     if (len(users_subscribed_with_old_email) == 1 and
-            users_subscribed_with_old_email[0].get_id == user.get_id):
-        safe_unsubscribe_user_from_mailchimp_list(
-            user,
-            list_id,
-            email=old_email,
-        )
-    safe_subscribe_user_to_mailchimp_list(
-        user,
-        list_id,
-        email=new_email,
-    )
+            users_subscribed_with_old_email[0] == user.get_id):
```

Comment:
I think there's an edge case here that won't be caught.  Suppose two users, user A and user B, have the same email address.  User A is subscribed only to commcare users, and user B is subscribed only to the mass email list.  If user B changes their email, the email address will not be unsubscribed from the mass email list even though now there does not exist any user with that email address who is subscribed to the mass email list (same issue happens if user A changed email address, only the commcare users list would be wrong).  To fix it, I think the subscribed() check and subsequent unsubscribe would have to happen separately for each list.